### ISSUE TITLE
Sphinx extension doesn't work when run outside the project root

### ISSUE THREAD

**Issue by apdavison** [fipymigrate/sumatra#168]
Michele Mattioni reported (see https://groups.google.com/forum/#!topic/sumatra-users/UFdbK5gDsJs):

"I store the image in an ad hoc file per run, however the Datastore path is store as relative, and when ran with sphinx the images and plot are not find.
 
The problem is that this line: 

https://bitbucket.org/apdavison/sumatra/src/658cec5709a0/src/publishing/utils.py?at=default#cl-86 

calls this line 

https://bitbucket.org/apdavison/sumatra/src/658cec5709a0/src/datastore/filesystem.py?at=default#cl-119 

which instantiate a new DataFile 
(https://bitbucket.org/apdavison/sumatra/src/658cec5709a0/src/datastore/filesystem.py?at=default#cl-16) 

The full path is determined combining the datastore with the path of the image. 
The problem I have is that, when I ran sphinx, the image is not found because is one directory too deep. 

Let me give you an example. My current structure is 
```
project_dir 
    .smt 
    doc 
    Results 
        20130325-133206 
        20130325-133234 
        20130325-135536 
   src 
```

The output are stored in the Results directory, more accurately in the 
subdirectories created from the timestamp. When I ran the make build 
from the doc directory, the datastore has as root 
"Results/20130325-133206" for example, e.g. 
```
    print (record.datastore) 
    Results/20130325-133206 
```
and the fullpath of the DataFile is than 
```
    self.full_path = os.path.join(store.root, path) 
#Results/20130325-133206, name_of_the_plot.png 
```

the problem is, sphinx is ran from the doc directory, therefore the 
real directory of the record_store is one level up, like 
`../Results/20130325-133206` 

I'm not sure how to fix this. TBH right now I have a terrible hack in 
place and this is like my get_image (in sumatra/publishing/utils) 
looks like 
```
    def get_image(record, image_path, sumatra_options, err=Exception): 

        if image_path is None: 
            image_key = record.output_data[0] 
        else: 
            image_key = None 
            for key in record.output_data: 
                if key.path == image_path: 
                    image_key = key 
                    break 
            if image_key is None: 
                raise ValueError("Record %s has no output data file 
with path %s. Valid paths are: %s" % ( 
                        record.label, image_path, ", ".join(key.path 
for key in record.output_data))) 
        assert isinstance(image_key, DataKey) 
        # check expected digest, if supplied, against key.digest 
        if ('digest' in sumatra_options 
            and sumatra_options['digest'] != image_key.digest): 
            raise err('Digests do not match') 
        record.datastore.root = os.path.join("../", 
record.datastore.root) #TERRIBLE HACK!! # We need a better way 
        return record.datastore.get_data_item(image_key)  # checks 
key.digest against file contents 
```

basically I overwrite the root of the datastore on the fly. But this 
works only if you have a project with my same structure. 
I'm not sure what is the best way to attack this." 

        

 _Imported from Bitbucket issue [182](https://bitbucket.org/apdavison/sumatra/issue/182),  created by apdavison on 05-30-2013 at 11:42, last modified: 04-16-2014 at 11:28_

**Comment by apdavison** [comment]
It does seem as though we need to store the absolute path for the datastore root, if we want to be able to work in subdirectories of the project root, or if we want to use the provenance data on a different machine to where it was captured. However, when accessing data stores in multiple ways (e.g. through symbolic links, NFS) it is very helpful to have the relative path (see ticket:2).

This suggests we need:

  1. a tool to re-map the roots for (some of?) the datastores in a separate copy of a given record store ("smt migrate"?)
  2. a tool to dynamically remap roots when accessing the same datastore through a symbolic link, NFS


        

_Trac comment by apdavison on 05-30-2013 at 21:31_
